contracts/TokenGovernance.sol

```diff
@@ -0,0 +1,149 @@
+    // make sure same proposals hasn't been opened before
+    require(proposals[_proposalHash].openTime == 0, "proposal already exists");
+    // get instance of token contract and pull proposal stake
+    leapToken.transferFrom(msg.sender, address(this), PROPOSAL_STAKE);
+    // create a new proposal in storage
+    proposals[_proposalHash] = Proposal(msg.sender, uint32(now), false, 0, 0);
+    // emit event for frontend
+    emit ProposalRegistered(_proposalHash, msg.sender);
+  }
+
+  function castVote(bytes32 _proposalHash, bytes32[] memory _proof, uint8 _outputIndex, bool isYes) public {
+    Proposal memory proposal = proposals[_proposalHash];
+    require(proposal.openTime > 0, "proposal does not exist");
+    require(proposal.finalized == false, "proposal already finalized");
+    uint32 timestamp;
+    (, timestamp,,) = bridge.periods(_proof[0]);
+    require(timestamp > 0, "The referenced period was not submitted to bridge");
+    require(timestamp <= proposal.openTime, "The transaction was submitted after the vote open time");
+    
+    bytes memory txData;
+    bytes32 txHash;
+    (, txHash, txData) = TxLib.validateProof(96, _proof);
+
+    // parse tx and check if it is usable for voting
+    TxLib.Tx memory tx = TxLib.parseTx(txData);
+    TxLib.Output memory out = tx.outs[_outputIndex];
+    require(out.value > 0, "UTXO has no value");
+    require(out.owner == msg.sender, "msg.sender not owner of utxo");
+    // TODO: fix
+    require(address(leapToken) == vault.getTokenAddr(out.color), "not Leap UTXO");
+
+    if (proposals[_proposalHash].votes[msg.sender] != 0) {
+      // TODO: clean up previous vote
+      _revertVote(proposals[_proposalHash], txHash);
+    }
+    if (isYes) {
+      proposals[_proposalHash].yesVotes += out.value;
+      proposals[_proposalHash].votes[msg.sender] = int256(out.value);
+    } else {
+      proposals[_proposalHash].noVotes += out.value;
+      proposals[_proposalHash].votes[msg.sender] = int256(out.value) * -1;
+    }
+    proposals[_proposalHash].usedTxns[txHash] = msg.sender;
+    emit VoiceCasted(_proposalHash, msg.sender, out.value, txHash);
+  }
+
+  function challengeUTXO(bytes32 _proposalHash, bytes32[] memory _proof, uint8 _inputIndex) public {
+    uint32 timestamp;
+    (, timestamp,,) = bridge.periods(_proof[0]);
+    require(timestamp > 0, "The referenced period was not submitted to bridge");
+    require(timestamp <= proposals[_proposalHash].openTime, "The transaction was submitted after the vote open time");
+    
+    bytes memory txData;
+    bytes32 txHash;
+    (, txHash, txData) = TxLib.validateProof(64, _proof);
+    // parse tx
+    TxLib.Tx memory txn = TxLib.parseTx(txData);
+    // checking that the transactions has been used in a vote
+    address signer = proposals[_proposalHash].usedTxns[txn.ins[_inputIndex].outpoint.hash];
+    // note: we don't check the output index here, assume that each transactions is only used for 1 vote
```

Comment:
I believe, you are reffering here that transaction can have more than a single output and - technically - one should be able to vote with all of them separetly?

Reply:
yes, exactly. i would like to move this to a later version. not sure it is worth covering the case now.